### PR TITLE
DOCS-#3954: Add Dask example notebooks

### LINKED ISSUE #3954: Add docs and notebook examples on running Modin with Dask
part of #3923

### PR DESCRIPTION
**Pull request by amyskov**
## What do these changes do?

This PR adds example Dask single node notebooks and corresponding tests.

- [ ] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [ ] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] signed commit with `git commit -s` 
- [x] Resolves #3954  
- [ ] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date 
- [ ] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

**Comment by YarShev** [comment]
This PR should be updated to be in line with the new structure introduced as part of #4214.

File: .github/workflows/ci-notebooks.yml [hunk 1]
```diff
@@ -24,15 +24,15 @@ jobs:
         with:
           path: taxi.csv
           # update cache only if notebooks require it to be changed
-          key: hashFiles("examples/tutorial/jupyter/execution/pandas_on_ray/local/**")
+          key: hashFiles("examples/tutorial/jupyter/**")
       # replace modin[all] with .[all] in the tutorial requirements file since we need
       # Modin built from sources
-      - run: sed -i 's/modin/./g' examples/tutorial/jupyter/execution/pandas_on_ray/requirements.txt
+      - run: sed -i 's/modin/./g' examples/tutorial/jupyter/requirements.txt
```
Comment (YarShev) [inline .github/workflows/ci-notebooks.yml:30]: We should have a single reqs file for each execution.

File: examples/tutorial/jupyter/test/test_notebooks.py [hunk 2]
```diff
@@ -92,29 +97,52 @@ def _replace_str(nb, original_str, str_to_replace):
     ].replace(original_str, str_to_replace)
 
 
+def _get_nb_dir(execution):
```
Comment (YarShev) [inline examples/tutorial/jupyter/test/test_notebooks.py:100]: Taking into account OmniSci we should have separate tests for each execution.

File: examples/tutorial/jupyter/README.md [hunk 3]
```diff
@@ -15,8 +15,10 @@ please install every package listed in `requirements.txt` file individually with
...
 
-A Jupyter Notebook example can be run from the current directory as follows:
+A Jupyter Notebook server example can be run from the current directory as follows:
 
 ```bash
-jupyter notebook local/exercise_1.ipynb
+jupyter notebook
 ```
+
+And then navigate to the needed notebook (for example to the `execution/pandas_on_ray/local/exercise_1.ipynb`).
```
Comment (amyskov) [inline examples/tutorial/jupyter/README.md:24]: Jupyter server should be started in the directory containing `img` directory, otherwise notebooks will not capable to access images.

File: .github/workflows/ci-notebooks.yml (hunk 1 above)
Comment (amyskov) [reply .github/workflows/ci-notebooks.yml]: I think you are right, there is no OmniSci package in pypi, so OmniSci notebooks need conda config. Separate reqs files are used.

File: examples/tutorial/jupyter/test/test_notebooks.py (hunk 2 above)
Comment (amyskov) [reply examples/tutorial/jupyter/test/test_notebooks.py]: But why? As for now OmniSci notebooks seems to me very similar to Ray/Dask notebooks. I think we can use the single test file as it is done now in this PR and avoid unnecessary duplication.

**Comment by amyskov** [comment]
@YarShev Dask and Ray notebooks have almost the same content, the only difference is that we should explicitly set `Engine` config when we use Dask. Shouldn't we combine Ray and Dask notebooks in the directory like `pandas_storage_format` and set optional cells with engine configuration? Such approach will help us to avoid code duplication and make notebooks more supportable.

File: examples/tutorial/jupyter/README.md (hunk 3 above)
Comment (YarShev) [reply examples/tutorial/jupyter/README.md:24]: Can we use absolute paths, but not relative?

File: examples/tutorial/jupyter/test/test_notebooks.py (hunk 2 above)
Comment (YarShev) [reply examples/tutorial/jupyter/test/test_notebooks.py]: Not sure OmniSci notebooks will seem to be very similar to Ray/Dask until #4001 is ready to merge. I would prefer separate them for now and combine when we have those in master as part of a new PR.

**Comment by YarShev** [comment]
> @YarShev Dask and Ray notebooks have almost the same content, the only difference is that we should explicitly set `Engine` config when we use Dask. Shouldn't we combine Ray and Dask notebooks in the directory like `pandas_storage_format` and set optional cells with engine configuration? Such approach will help us to avoid code duplication and make notebooks more supportable.

Let's allow users to use explicit notebooks. That is more user-friendly instead of providing optional cells.

File: examples/tutorial/jupyter/test/test_notebooks.py (hunk 2 above)
Comment (amyskov) [reply examples/tutorial/jupyter/test/test_notebooks.py]: Ok, let's use unique tests for each of the executions.

File: examples/tutorial/jupyter/README.md (hunk 3 above)
Comment (amyskov) [reply examples/tutorial/jupyter/README.md]: Fixed.

File: examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+ep = ExecutePreprocessor(timeout=600, kernel_name="python3")
+
+download_taxi_dataset = f"""import os
+import urllib.request
+if not os.path.exists("{test_dataset_path}"):
+    url_path = "https://s3.amazonaws.com/nyc-tlc/trip+data/yellow_tripdata_2021-01.csv"
+    urllib.request.urlretrieve(url_path, "{test_dataset_path}")
+    """
+
+notebooks_dir = "examples/tutorial/jupyter/execution/pandas_on_dask/local"
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py:29]:
```suggestion
local_notebooks_dir = "examples/tutorial/jupyter/execution/pandas_on_dask/local"
```

File: examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+
+
+def _execute_notebook(notebook):
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py:32]: See the comment about moving common stuff.

File: examples/tutorial/jupyter/execution/pandas_on_dask/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,7 @@
+FROM continuumio/miniconda3
+
+COPY ./requirements.txt ./requirements.txt
+
+RUN conda install -c conda-forge psutil setproctitle
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/Dockerfile:5]: See the related comment.

File: examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_1.ipynb [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Concept for exercise: setting Modin engine\n",
+    "\n",
+    "Modin uses Ray engine by default so no additional action is required to start to use it. Alternatively, if you need to use another engine, it should be specified either by setting the Modin config or by setting environment variable before the first `DataFrame` creation or data ingestion as it shown below. Also note, that full list of Modin configs and corresponding environment variables can be found in the [special section](https://modin.readthedocs.io/en/stable/flow/modin/config.html#modin-configs-list) of the Usage Guide in the Modin documentation."
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_1.ipynb:65]:
```suggestion
    "Modin uses Ray as an execution engine by default so no additional action is required to start to use it. Alternatively, if you need to use another engine, it should be specified either by setting the Modin config or by setting Modin environment variable before the first operation with Modin as it is shown below. Also, note that the full list of Modin configs and corresponding environment variables can be found in the [Modin Configuration Settings](https://modin.readthedocs.io/en/stable/flow/modin/config.html#modin-configs-list) section of the Modin documentation."
```

File: examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_1.ipynb [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+    "df = pd.DataFrame(frame_data)\n",
+    "```\n",
+    "\n",
+    "When creating a dataframe from a non-distributed object, it will take extra time to partition the data. When this is happening, you will see this message:\n",
+    "\n",
+    "```\n",
+    "UserWarning: Distributing <class 'numpy.ndarray'> object. This may take some time.\n",
+    "```\n",
+    "\n",
+    "**In a later exercise, we will discuss times when it is not possible to speed up the computation, even with multiprocessing or multithreading.**"
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_1.ipynb:105]:
```suggestion
```
There is no any use of multiprocessing or multithreading in a later exercise so this should be removed.

File: examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_1.ipynb [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Modin engine can be specified either by config\n",
+    "import modin.config as cfg\n",
+    "cfg.Engine.put(\"dask\")\n",
+    "\n",
+    "# or environment variable setting\n",
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_1.ipynb:78]:
```suggestion
    "# or by setting the environment variable\n",
```

File: examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_1.ipynb [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    "# os.environ[\"MODIN_ENGINE\"] = \"dask\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Concept for exercise: Dataframe constructor\n",
+    "\n",
+    "Often when playing around in Pandas, it is useful to create a DataFrame with the constructor. That is where we will start.\n",
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_1.ipynb:89]:
```suggestion
    "Often when playing around in pandas, it is useful to create a DataFrame with the constructor. That is where we will start.\n",
```
Pandas -> pandas, everywhere

File: examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_1.ipynb [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Note: Do not change this code!\n",
+    "import numpy as np\n",
+    "import pandas\n",
+    "import subprocess\n",
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_1.ipynb:117]:
```suggestion
```
Nowhere used.

File: examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_1.ipynb [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+   "outputs": [],
+   "source": [
+    "df.count()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Please move on to Exercise 2 when you are ready**"
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_1.ipynb:231]:
```suggestion
    "**Please move on to [Exercise 2](./exercise_2.ipynb) when you are ready**"
```
everywhere

File: examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_2.ipynb [hunk 13]
```diff
@@ -0,0 +1,495 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import modin.pandas as pd\n",
+    "import pandas as old_pd\n",
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_2.ipynb:48]:
```suggestion
    "import pandas\n",
```
everywhere

File: .github/workflows/ci-notebooks.yml [hunk 14]
```diff
@@ -11,6 +11,9 @@ jobs:
   test-notebooks:
```
Comment (YarShev) [inline .github/workflows/ci-notebooks.yml:11]:
```suggestion
  tutorial-notebooks:
```

File: .github/workflows/ci-notebooks.yml [hunk 15]
```diff
@@ -11,6 +11,9 @@ jobs:
...
     name: test notebooks
```
Comment (YarShev) [inline .github/workflows/ci-notebooks.yml:12]:
```suggestion
    name: tutorial notebooks
```

File: examples/tutorial/jupyter/README.md (hunk 3 above)
Comment (YarShev) [reply examples/tutorial/jupyter/README.md]: I see the relative paths. Can we use absolute paths instead?

Comment (amyskov) [reply examples/tutorial/jupyter/README.md]: Absolute patrh `examples/tutorial/jupyter/execution/pandas_on_ray/local/exercise_1.ipynb` is used now.

File: examples/tutorial/jupyter/execution/pandas_on_ray/Dockerfile [hunk 16]
```diff
@@ -1,5 +1,7 @@
 FROM continuumio/miniconda3
 
+COPY ./requirements.txt ./requirements.txt
```
[the remark this answers is not in the document]
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_ray/Dockerfile:3]: We need to copy requirement file into the container to use it with pip afterwards (pip packages installation will be done inside the container and we need to copy all needed data explicitly).

File: examples/tutorial/jupyter/execution/pandas_on_ray/Dockerfile [hunk 17]
```diff
@@ -1,5 +1,7 @@
...
+
 RUN conda install -c conda-forge psutil setproctitle
```
[the remark this answers is not in the document]
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_ray/Dockerfile:5]: I believe we can drop it, but unfortunately i don't have access to `docker` now to test it...

File: examples/tutorial/jupyter/execution/pandas_on_ray/test/test_notebooks.py [hunk 18]
```diff
@@ -24,6 +26,8 @@
     urllib.request.urlretrieve(url_path, "{test_dataset_path}")
     """
 
+notebooks_dir = "examples/tutorial/jupyter/execution/pandas_on_ray/local"
+
 
 def _execute_notebook(notebook):
```
[the remark this answers is not in the document]
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_ray/test/test_notebooks.py]: Ok, utils were moved to `examples/tutorial/jupyter/execution/test/utils.py`.

File: examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py (hunk 5 above)
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py]: Moved.

File: examples/tutorial/jupyter/execution/pandas_on_dask/Dockerfile [hunk 19]
```diff
@@ -0,0 +1,7 @@
+FROM continuumio/miniconda3
+
+COPY ./requirements.txt ./requirements.txt
```
[the remark this answers is not in the document]
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_dask/Dockerfile:3]: Answered here https://github.com/modin-project/modin/pull/4139#discussion_r813956900.

File: examples/tutorial/jupyter/execution/pandas_on_dask/Dockerfile (hunk 6 above)
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_dask/Dockerfile:5]: Answered here https://github.com/modin-project/modin/pull/4139#discussion_r813958147.

File: examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_2.ipynb [hunk 20]
```diff
@@ -0,0 +1,495 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Concept for Exercise: `read_csv` speedups\n",
+    "\n",
+    "The most commonly used data ingestion method used in pandas is CSV files (link to pandas survey). This concept is designed to give an idea of the kinds of speedups possible, even on a non-distributed filesystem. Modin also supports other file formats for parallel and distributed reads, which can be found in the documentation.\n",
+    "\n",
+    "![](../../../img/read_csv_perf.png)"
```
[the remark this answers is not in the document]
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_2.ipynb]: Ok, graph was removed.

File: examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_2.ipynb (hunk 13 above)
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_2.ipynb]: Fixed.

File: examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_4.ipynb [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+    "\n",
+    "**GOAL**: Explore some of the experimental features being added to Modin."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "9b487c51",
+   "metadata": {},
+   "source": [
+    "### Concept for exercise: Progress Bar\n",
```
[the remark this answers is not in the document]
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_4.ipynb]: Right, section with progress bar was removed.

File: .github/workflows/ci-notebooks.yml (hunk 14 above)
Comment (YarShev) [reply .github/workflows/ci-notebooks.yml:11]: Why wasn't this applied?

File: .github/workflows/ci-notebooks.yml (hunk 15 above)
Comment (YarShev) [reply .github/workflows/ci-notebooks.yml:12]: Why wasn't this applied?

File: examples/tutorial/jupyter/README.md [hunk 22]
```diff
@@ -15,8 +15,10 @@ please install every package listed in `requirements.txt` file individually with
 
 ## Run Jupyter Notebooks
 
-A Jupyter Notebook example can be run from the current directory as follows:
+A Jupyter Notebook server example can be run from the current directory as follows:
```
Comment (YarShev) [inline examples/tutorial/jupyter/README.md:18]:
```suggestion
A Jupyter Notebook server can be run from the current directory as follows:
```
?

File: examples/tutorial/jupyter/README.md [hunk 23]
```diff
@@ -15,8 +15,10 @@ please install every package listed in `requirements.txt` file individually with
 
 ## Run Jupyter Notebooks
```
Comment (YarShev) [inline examples/tutorial/jupyter/README.md:16]: Why was this file moved? We should have a single one for pandas_on_ray and another one for pandas_on_dask in the respective directories. Remember about the OmniSci engine.

File: examples/tutorial/jupyter/README.md [hunk 24]
```diff
@@ -15,8 +15,10 @@ please install every package listed in `requirements.txt` file individually with
...
 
-A Jupyter Notebook example can be run from the current directory as follows:
+A Jupyter Notebook server example can be run from the current directory as follows:
 
 ```bash
-jupyter notebook local/exercise_1.ipynb
+jupyter notebook
 ```
+
+And then navigate to the needed notebook (for example to the `examples/tutorial/jupyter/execution/pandas_on_ray/local/exercise_1.ipynb`).
```
Comment (YarShev) [inline examples/tutorial/jupyter/README.md:24]:
```suggestion
Navigate to a concrete notebook (for example to the `examples/tutorial/jupyter/execution/pandas_on_ray/local/exercise_1.ipynb`).
```

File: examples/tutorial/jupyter/execution/pandas_on_ray/Dockerfile (hunk 16 above)
[the remark this answers is not in the document]
Comment (YarShev) [reply examples/tutorial/jupyter/execution/pandas_on_ray/Dockerfile:3]: ok

File: examples/tutorial/jupyter/execution/pandas_on_ray/Dockerfile (hunk 17 above)
[the remark this answers is not in the document]
Comment (YarShev) [reply examples/tutorial/jupyter/execution/pandas_on_ray/Dockerfile:5]: Could you create an issue to resolve this as part of a new PR?

File: examples/tutorial/jupyter/execution/pandas_on_ray/test/test_notebooks.py [hunk 25]
```diff
@@ -11,97 +11,35 @@
...
-    import_cell_idx = [
-        idx
-        for idx, cell in enumerate(nb["cells"])
-        if cell["cell_type"] == "code" and identifier in cell["source"]
-    ]
-    assert len(import_cell_idx) == 1
-    return import_cell_idx[0]
+import nbformat
 
+UTILS_PARENT_DIR = os.path.abspath(
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_ray/test/test_notebooks.py:19]: Can we just use the following import below to import required stuff?
```python
from modin.examples.tutorial.jupyter.execution.test.utils import ...
```

File: examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+import os
+import sys
+
+import nbformat
+
+UTILS_PARENT_DIR = os.path.abspath(
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py:19]: See the related comment.

File: examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+    os.path.join(os.path.dirname(__file__), *[".." for _ in range(2)])
+)
+sys.path.insert(0, UTILS_PARENT_DIR)
+from test.utils import (  # noqa
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py:23]: Why noqa?

**Review by YarShev — COMMENTED** [review]
Looks pretty good. Some more a little efforts.

File: examples/tutorial/jupyter/execution/pandas_on_ray/Dockerfile (hunk 17 above)
[the remark this answers is not in the document]
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_ray/Dockerfile]: Ok, issue was created https://github.com/modin-project/modin/issues/4267.

File: .github/workflows/ci-notebooks.yml (hunk 14 above)
Comment (amyskov) [reply .github/workflows/ci-notebooks.yml]: It wasn't applied by mistake, so it is applied now.

File: .github/workflows/ci-notebooks.yml (hunk 15 above)
Comment (amyskov) [reply .github/workflows/ci-notebooks.yml]: Applied now.

File: examples/tutorial/jupyter/README.md (hunk 23 above)
Comment (amyskov) [reply examples/tutorial/jupyter/README.md:16]: It was moved to this directory to make it more convenient to user where to start jupyter notebooks - we should start notebooks in the directory where images are located and if we move readme closer to notebooks we will need to tell something like "go in the parent directory, then once again and only then you should start jupyter server". Additionally, if we place readme here we will be capable to avoid bunch of code duplication i believe (for OmniSci specific notes separate section can be created).

File: examples/tutorial/jupyter/execution/pandas_on_ray/test/test_notebooks.py (hunk 25 above)
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_ray/test/test_notebooks.py:19]: We can make import relative to modin root directory, but in this case, we will need to put `__init__.py` in each nested directory (in `examples`, `examples/tutorial`, `examples/tutorial/jupyter`, `examples/tutorial/jupyter/execution`), otherwise we won't be capable to access these directories as python modules.

Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_ray/test/test_notebooks.py]: Ah, wait. Actually we can do it, but wee need to tune inserted path too. Import path relative to Modin root directory was used.

File: examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py (hunk 26 above)
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py]: Fixed.

File: examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py (hunk 27 above)
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py]: `noqa` is needed to avoid `E402 module level import not at top of file` flake8 error since we are using `sys.path.insert` prior of utils import.

Comment (YarShev) [reply examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py]: I see. Let's use `noqa: E402` explicitly then.

Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py]: Fixed.

File: examples/tutorial/jupyter/README.md (hunk 23 above)
Comment (YarShev) [reply examples/tutorial/jupyter/README.md:16]: This file should be corrected with respect to different exexutions.

File: examples/tutorial/jupyter/README.md [hunk 28]
```diff
@@ -15,8 +15,10 @@ please install every package listed in `requirements.txt` file individually with
...
 
-A Jupyter Notebook example can be run from the current directory as follows:
+A Jupyter Notebook server can be run from the current directory as follows:
 
 ```bash
-jupyter notebook local/exercise_1.ipynb
+jupyter notebook
 ```
+
+Navigate to a concrete notebook (for example to the `examples/tutorial/jupyter/execution/pandas_on_ray/local/exercise_1.ipynb`).
```
Comment (YarShev) [inline examples/tutorial/jupyter/README.md:24]:
```suggestion
Navigate to a concrete notebook (for example, to the `execution/pandas_on_ray/local/exercise_1.ipynb`).
```
Since a jupyter server is run from the current directory we should use relative paths to notebooks.

File: examples/tutorial/jupyter/execution/pandas_on_dask/requirements.txt [hunk 29]
```diff
@@ -0,0 +1,7 @@
+fsspec
+s3fs
+jupyterlab
+ipywidgets
+tqdm
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/requirements.txt:5]:
```suggestion
```
I believe this should be removed since it is related to the progress bar feature.

File: examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+import os
+import sys
+
+import nbformat
+
+MODIN_DIR = os.path.abspath(
```
Comment (YarShev) [inline examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py:19]: What about just using this?
```python
import modin
modin.__file__
```

File: docs/getting_started/examples.rst [hunk 31]
```diff
@@ -17,23 +17,17 @@ The following notebooks demonstrate how Modin can be used for scalable data scie
...
-- Exercise 6: Running Modin in a Cluster Environment [`Source <https://github.com/modin-project/modin/blob/721d2d2fe2a46d3b7442e1ab2c7b70ad3bff0712/examples/tutorial/jupyter/execution/pandas_on_ray/cluster/exercise_6.ipynb>`_]
+- Exercise 4: Experimental Features in Modin (Spreadsheet, Progress Bar) [`Source PandasOnRay <https://github.com/modin-project/modin/blob/721d2d2fe2a46d3b7442e1ab2c7b70ad3bff0712/examples/tutorial/jupyter/execution/pandas_on_ray/local/exercise_4.ipynb>`_, `Source PandasOnDask <https://github.com/modin-project/modin/blob/721d2d2fe2a46d3b7442e1ab2c7b70ad3bff0712/examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_4.ipynb>`_]
+- Exercise 5: Setting up Modin in a Cluster Environment [`Source PandasOnRay <https://github.com/modin-project/modin/blob/721d2d2fe2a46d3b7442e1ab2c7b70ad3bff0712/examples/tutorial/jupyter/execution/pandas_on_ray/cluster/exercise_5.ipynb>`_]
+- Exercise 6: Running Modin in a Cluster Environment [`Source PandasOnRay <https://github.com/modin-project/modin/blob/721d2d2fe2a46d3b7442e1ab2c7b70ad3bff0712/examples/tutorial/jupyter/execution/pandas_on_ray/cluster/exercise_6.ipynb>`_]
```
Comment (YarShev) [inline docs/getting_started/examples.rst:30]:
```suggestion
- Exercise 6: Running Modin in a Cluster Environment [`Source PandasOnRay <https://github.com/modin-project/modin/blob/721d2d2fe2a46d3b7442e1ab2c7b70ad3bff0712/examples/tutorial/jupyter/execution/pandas_on_ray/cluster/exercise_6.ipynb>`_]

How to get required dependencies for the tutorial notebooks and to run them please refer to the respective [README.md](examples/tutorial/jupyter/README.md) file.
```
Please correct the link.

File: docs/usage_guide/examples/index.rst [hunk 32]
```diff
@@ -7,23 +7,17 @@ Modin in the cloud, the use of Modin spreadsheet.
...
-- Exercise 6: Running Modin in a Cluster Environment [`Source <https://github.com/modin-project/modin/blob/721d2d2fe2a46d3b7442e1ab2c7b70ad3bff0712/examples/tutorial/jupyter/execution/pandas_on_ray/cluster/exercise_6.ipynb>`_]
+- Exercise 4: Experimental Features in Modin (Spreadsheet, Progress Bar) [`Source PandasOnRay <https://github.com/modin-project/modin/blob/721d2d2fe2a46d3b7442e1ab2c7b70ad3bff0712/examples/tutorial/jupyter/execution/pandas_on_ray/local/exercise_4.ipynb>`_, `Source PandasOnDask <https://github.com/modin-project/modin/blob/721d2d2fe2a46d3b7442e1ab2c7b70ad3bff0712/examples/tutorial/jupyter/execution/pandas_on_dask/local/exercise_4.ipynb>`_]
+- Exercise 5: Setting up Modin in a Cluster Environment [`Source PandasOnRay <https://github.com/modin-project/modin/blob/721d2d2fe2a46d3b7442e1ab2c7b70ad3bff0712/examples/tutorial/jupyter/execution/pandas_on_ray/cluster/exercise_5.ipynb>`_]
+- Exercise 6: Running Modin in a Cluster Environment [`Source PandasOnRay <https://github.com/modin-project/modin/blob/721d2d2fe2a46d3b7442e1ab2c7b70ad3bff0712/examples/tutorial/jupyter/execution/pandas_on_ray/cluster/exercise_6.ipynb>`_]
```
Comment (YarShev) [inline docs/usage_guide/examples/index.rst:20]:
```suggestion
- Exercise 6: Running Modin in a Cluster Environment [`Source PandasOnRay <https://github.com/modin-project/modin/blob/721d2d2fe2a46d3b7442e1ab2c7b70ad3bff0712/examples/tutorial/jupyter/execution/pandas_on_ray/cluster/exercise_6.ipynb>`_]

How to get required dependencies for the tutorial notebooks and to run them please refer to the respective [README.md](examples/tutorial/jupyter/README.md) file.
```
Please correct the link.

File: examples/tutorial/jupyter/README.md (hunk 23 above)
Comment (amyskov) [reply examples/tutorial/jupyter/README.md:24]: Right, notes for different executions are added.

File: docs/getting_started/examples.rst (hunk 31 above)
Comment (amyskov) [reply docs/getting_started/examples.rst]: Link was corrected.

File: docs/usage_guide/examples/index.rst (hunk 32 above)
Comment (amyskov) [reply docs/usage_guide/examples/index.rst]: Link was corrected.

File: examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py (hunk 30 above)
Comment (amyskov) [reply examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py:19]: I think usage of this approach can bring us additional complication - this approach strongly requires modin built from sources and only in development mode (pip command should be `pip install -e .[dask/ray]`).

File: examples/tutorial/jupyter/README.md [hunk 33]
```diff
@@ -0,0 +1,32 @@
+# Jupyter notebook examples to run Modin on PandasOnRay or PandasOndask
+
```
Comment (YarShev) [inline examples/tutorial/jupyter/README.md:2]:
```suggestion
# Jupyter notebook examples to run with Modin

Currently we provide tutorial notebooks for the following execution backends:

- [PandasOnRay](https://modin.readthedocs.io/en/latest/development/using_pandas_on_ray.html)
- [PandasOnDask](https://modin.readthedocs.io/en/latest/development/using_pandas_on_dask.html)

```

File: examples/tutorial/jupyter/README.md [hunk 34]
```diff
@@ -0,0 +1,32 @@
...
+To get required dependencies for these Jupyter Notebooks
+you should create a development environment with `pip`
+using `requirements.txt` file located in the respective directory:
+
+```bash
+pip install -r execution/pandas_on_ray/requirements.txt
+```
+
+to install dependencies needed to run notebooks using Ray engine or
```
Comment (YarShev) [inline examples/tutorial/jupyter/README.md:13]:
```suggestion
to install dependencies needed to run notebooks with Modin on `PandasOnRay` execution or
```

File: examples/tutorial/jupyter/README.md [hunk 35]
```diff
@@ -0,0 +1,32 @@
...
+
+```bash
+pip install -r execution/pandas_on_dask/requirements.txt
+```
+
+to install dependencies needed to run notebooks using Dask engine
```
Comment (YarShev) [inline examples/tutorial/jupyter/README.md:19]:
```suggestion
to install dependencies needed to run notebooks with Modin on `PandasOnDask` execution.
```

File: examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py (hunk 30 above)
Comment (YarShev) [reply examples/tutorial/jupyter/execution/pandas_on_dask/test/test_notebooks.py:19]: I see, makes sense.

**Review by YarShev — APPROVED** [review]
@amyskov, LGTM, thanks!

**Comment by Garra1980** [comment]
> @amyskov, LGTM, thanks!

👍👍👍
